crates/modules/network-monitor/probes.bpf.c

```diff
@@ -444,172 +482,202 @@ int BPF_PROG(sys_exit_accept, struct pt_regs *regs, int __syscall_nr,
+        break;
+      }
+      break;
+    case ETH_P_IPV6:
+      switch (direction) {
+      case EGRESS:
+        copy_ipv6_tcphdr_source(th, &msg_event->source);
+        copy_ipv6_tcphdr_dest(th, &msg_event->destination);
+        break;
+      case INGRESS:
+        copy_ipv6_tcphdr_source(th, &msg_event->destination);
+        copy_ipv6_tcphdr_dest(th, &msg_event->source);
+        break;
       }
-    } else {
-      event->send.data.len = 0;
-
-      copy_skc_dest(&sk->__sk_common, &event->recv.destination);
     }
-  } else if (skb->protocol == bpf_htons(ETH_P_IPV6)) {
-    if (data + sizeof(struct ipv6hdr) > data_end)
-      goto output;
+    break;
+  }
+  case IPPROTO_UDP: {
+    if (data + headers_len + sizeof(struct udphdr) > data_end) {
+      LOG_ERROR("found a UDP packet too small to fit a UDP header");
+      goto pass;
+    }
 
-    struct ipv6hdr *ih6 = data;
-    event->send.proto = ih6->nexthdr;
-
-    if (ih6->nexthdr == IPPROTO_UDP) {
-      __u32 headers_len = sizeof(struct ipv6hdr) + sizeof(struct udphdr);
-
-      event->send.data_len = skb->len - headers_len;
-      // read data
-      read_sk_buff(&event->buffer, &event->send, skb, headers_len);
-
-      // in UDP we find destination value in sockaddr
-      // NOTE: msg_name is NULL if the userspace code is not interested
-      // in knowing the source of the message. In that case we won't extract
-      // the source port and address.
-      if (args){
-        struct sockaddr *addr = args->data[2];
-        if (!addr) {
-          LOG_DEBUG("sockaddr is null. ");
-        } else {
-          copy_sockaddr(addr, &event->recv.destination, true);
-        }
+    struct udphdr *uh = data + headers_len;
+    headers_len += sizeof(struct udphdr);
+
+    switch (l3_proto) {
+    case ETH_P_IPV4:
+      switch (direction) {
+      case EGRESS:
+        copy_ipv4_udphdr_source(uh, &msg_event->source);
+        copy_ipv4_udphdr_dest(uh, &msg_event->destination);
+        break;
+      case INGRESS:
+        copy_ipv4_udphdr_source(uh, &msg_event->destination);
+        copy_ipv4_udphdr_dest(uh, &msg_event->source);
+        break;
       }
-    } else {
-      event->send.data.len = 0;
-
-      copy_skc_dest(&sk->__sk_common, &event->recv.destination);
+      break;
+    case ETH_P_IPV6:
+      switch (direction) {
+      case EGRESS:
+        copy_ipv6_udphdr_source(uh, &msg_event->source);
+        copy_ipv6_udphdr_dest(uh, &msg_event->destination);
+        break;
+      case INGRESS:
+        copy_ipv6_udphdr_source(uh, &msg_event->destination);
+        copy_ipv6_udphdr_dest(uh, &msg_event->source);
+        break;
+      }
+      break;
     }
+
+    buffer_index_init(&network_event->buffer, &msg_event->data);
+    buffer_append_skb_bytes(&network_event->buffer, &msg_event->data, skb, headers_len);
+    break;
   }
+  default:
+    LOG_DEBUG("ignored unsupported L4 protocol %d", l4_proto);
+    goto send_event;
+  }
+
+  msg_event->data_len = skb->len - headers_len;
```

Comment:
there is no way  to catch errors from `buffer_append_skb_bytes` at this level.

in case of errors
```c
msg_event->data_len = skb->len - headers_len;
```
is always executed, even if there are conditions not met on the same fields inside `buffer_append_skb_bytes`:
```c
  if (offset >= skb->len) {
    LOG_ERROR("invalid offset: %zu, greater than packet length: %zu", offset, skb->len);
    return;
  }
```

Probably it doesn't create any issues most of the time, but it seems strange

Reply:
My reasoning was that even if we fail to read the packet content (for whatever reason), we still want to return the correct information about the length of the packet.

I was thinking about not emitting the event in general in such case, but the thing is that we don't include the packet content for the most of events anyways. We include the content only for UDP. For TCP, we return the event which has `data_len` included, but no payload. So I feel like that abruptly stopping the event emission isn't quite right.

Reply:
But hmm... it's probably cleaner if I return the error code `-1` in `buffer_append_skb_bytes` and then explicitly handle it by doing something like

```c
if (buffer_append_skb_bytes(...) < 0) {
    LOG_ERROR("failed to get the packet payload");
}
```

To make that explicity and keep it equivalent to handling expected errors with `if Err(e)` in Rust.

Reply:
Done